src/distributed-transaction/percolator.md

```diff
@@ -0,0 +1,150 @@
+
+### Writing
+
+Percolator's transactions are committed by a 2PC algorithm. Its two phases are called `Prewrite` and `Commit`.
+
+In `Prewrite` phase:
+1. We get a timestamp from the timestamp oracle, and we call the timestamp the transaction's `start_ts`.
+2. For each row involved in the transaction, put a lock to `lock` column. At the same time, put the value to the `data` column with the timestamp `start_ts`. One of these locks will be chosen as the *primary* lock, and others are *secondary* locks. Each lock saves the transaction's `start_ts`. The secondary locks, in addition, save the primary lock's location.
+    * If we found that there's already a lock or newer version than `start_ts`, the current transaction will be rolled back.
+
+And then, in `Commit` phase:
+1. Get another timestamp, namely `commit_ts`.
+2. Remove the primary lock, and at the same time write a record to `write` column with the timestamp `commit_ts`, and its value records the transaction's `start_ts`.
+    * If the lock is missing, the commit should fail.
+3. For all secondaries, do the same thing.
+
+Once the step 2 (committing the primary) is done, we say the whole transaction is done. It doesn't matter if the process of committing the secondaries unexpectedly exited. When we meet a lock of other transactions, we can see whether its primary has been committed to determine whether the whole transaction has been committed.
+
+Let's see the example in the paper of Percolator. Assume we are writing two rows in a single transaction. At first, the data looks like this:
+
+![Figure 1](percolator-figure-1.png)
+
+This table shows Bob and Joe's balance. Now Bob want to transfer his $7 to Joe's account. First, we do `Prewrite`:
+
+1. Get the `start_ts` of the transaction. We got `7` in this example.
+2. For each row involved in this transaction, put a lock in the `lock` column, and put the data to the `data` column. One of the locks will be chose as the primary lock.
+
+After `Prewrite`, our data looks like this:
+
+![Figure 2](percolator-figure-2.png)
+
+Then `Commit`:
+
+1. Get the `commit_ts`. We got `8`.
+2. Commit the primary: Remove the primary lock and put the commit record to `write` column. We get this:
+    ![Figure 3](percolator-figure-3.png)
+3. Commit all secondaries. Finally, we get this:
+    ![Figure 4](percolator-figure-4.png)
+
+### Reading
+
+To read from Percolator, a timestamp is required too. The procedure to perform read operation is as following:
+
+1. Get a timestamp, `ts`.
+2. Check if the row we are going to read is locked with timestamp in the range `[0, ts]`;
+    * If so, it means the row is locked by a earlier-started transaction. Then we are not sure whether that transaction will be committed before `ts` or after `ts`. Other wise, we can continue reading.
```

Comment:
```suggestion
    * If so, it means the row is locked by an earlier-started transaction. Then we are not sure whether that transaction will be committed before `ts` or after `ts`. Otherwise, we can continue reading.
```